crates/tick/src/clock_control.rs

```diff
@@ -0,0 +1,819 @@
+///
+/// let now = clock.system_time();
+///
+/// assert_eq!(
+///     clock.system_time().duration_since(now)?,
+///     Duration::from_secs(1)
+/// );
+///
+/// # Ok::<(), Box<dyn std::error::Error>>(())
+/// ```
+///
+/// # Production code and `ClockControl`
+///
+/// You should **never** enable the `test-util` feature or use [`ClockControl`] in production code.
+/// When the `test-util` feature is enabled, extra code is compiled into the binary to support
+/// testing scenarios. This extra code hampers performance when running in production.
+///
+/// Always ensure that the `test-util` feature is only enabled for `dev-dependencies`.
+///
+/// ```toml
+/// tick = { version = "*", features = ["test-util"] }
+/// ```
+#[derive(Debug, Clone, Default)]
+pub struct ClockControl {
+    /// Clock control requires controlling the passage of time across threads.
+    /// For this reason, we need to use a mutex to ensure that state is consistent
+    /// across all threads.
+    state: Arc<Mutex<State>>,
+}
+
+impl ClockControl {
+    /// Creates a new `ClockControl` instance.
+    ///
+    /// By default, auto-advance is disabled and the initial time is set to the UNIX epoch.
+    ///
+    /// # Examples
+    /// ```
+    /// use std::time::{Duration, SystemTime};
+    ///
+    /// use tick::ClockControl;
+    ///
+    /// let clock = ClockControl::new()
+    ///     .auto_advance(Duration::from_secs(1))
+    ///     .to_clock();
+    ///
+    /// let time1 = clock.system_time();
+    /// let time2 = clock.system_time();
+    ///
+    /// assert_eq!(
+    ///     time2.duration_since(time1)?,
+    ///     Duration::from_secs(1)
+    /// );
+    ///
+    /// # Ok::<(), Box<dyn std::error::Error>>(())
+    /// ```
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            state: Arc::new(Mutex::new(State::default())),
+        }
+    }
+
+    /// Creates a new `ClockControl` instance at the specified time.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::time::{Duration, SystemTime};
+    ///
+    /// use tick::ClockControl;
+    ///
+    /// // Create clock at a specific system time
+    /// let system_time = SystemTime::UNIX_EPOCH + Duration::from_secs(100);
+    /// let control = ClockControl::new_at(system_time);
+    /// let clock = control.to_clock();
+    /// assert_eq!(clock.system_time(), system_time);
+    /// ```
+    #[must_use]
+    pub fn new_at(time: impl Into<SystemTime>) -> Self {
+        let this = Self::new();
+        this.advance_to(time.into());
+        this
+    }
+
+    /// Creates a new `ClockControl` instance with the current system time.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use std::time::SystemTime;
+    ///
+    /// use tick::ClockControl;
+    ///
+    /// let control = ClockControl::now();
+    /// let clock = control.to_clock();
+    ///
+    /// assert!(SystemTime::now() >= clock.system_time());
+    /// ```
+    #[must_use]
+    pub fn now() -> Self {
```

Comment:
What's the use case for this? Feels like it is inviting users to mix "real-time" with "fake-time".

Reply:
dropped, it's indeed not required